File: client/library/java/ihart/CVManager.java
package ihart;

import java.net.*;
import java.util.ArrayList;
import java.util.List;
import java.io.*;
import javax.json.Json;
import javax.json.JsonArray;
import javax.json.JsonObject;
import javax.json.JsonReader;
import javax.json.JsonValue;
import javax.swing.event.EventListenerList;

/**
 * CVManager is an event dispatcher that uses a socket's dataHandler to alert
 * the user of new events.
 **/

public class CVManager {
	/**
	 * Fields
	 */
	private String hostname = "localhost";
	private int port;

	protected EventListenerList listenerList = new EventListenerList();

	/**
	 * The constructor sets up the socket with the given hostName and port, then
	 * connects and sets up listeners for data received and dispatches
	 * corresponding CVEvents when motion events occur.
	 * 
	 * @param hostname
	 *            The host to connect to; if the same computer is running the
	 *            server and the application, use "localhost".
	 * @param port
	 *            The port number to connect to. The default for the server is
	 *            5204; use this unless you know the server is running on a
	 *            different port.
	 **/
	public CVManager(String hostname, int port) {
		this.hostname = hostname;
		this.port = port;
		start();
	}

	/**
	 * Create a thread that runs the dataHandler for the port.
	 */
	private void start() {
		new Thread(new Runnable() {
			public void run() {
				dataHandler(port);
			}
		}).start();
	}

	/**
	 * Add a listener to the Event Listener List.
	 * 
	 * @param listener
	 *            The listener to be added to the list
	 */
	public void addCVEventListener(CVEventListener listener) {
		listenerList.add(CVEventListener.class, listener);
	}

	/**
	 * Remove a listener from the Event Listener List.
	 * 
	 * @param listener
	 *            The listener to be removed from the list
	 */
	public void removeCVEventListener(CVEventListener listener) {
		listenerList.remove(CVEventListener.class, listener);
	}

	/**
	 * Parse the data from the server, store the data, and dispatch the
	 * appropriate events.
	 * 
	 * @param port
	 *            The to connect to
	 */
	private void dataHandler(int port) {
		// the string of data from the server
		String fromServer;

		Socket mySocket = null;
		PrintWriter out = null;
		BufferedReader in = null;

		try {
			// connect to the server with host name "localhost" on port 5204
			mySocket = new Socket("localhost", port);
			out = new PrintWriter(mySocket.getOutputStream(), true);
			in = new BufferedReader(new InputStreamReader(mySocket.getInputStream(), "UTF-8"));
			System.out.println("Connected to localhost");

			while ((fromServer = in.readLine()) != null) {
				try {
					// Remove newline/carriage returns because the JSON parser
					// doesn't like them.
					fromServer = fromServer.replaceAll("\n", "");
					fromServer = fromServer.replaceAll("\0", "");
					fromServer = fromServer.replaceAll("\r", "");

					JsonReader reader = Json.createReader(new StringReader(fromServer));

					// We start with an array of the areas of interest.
					JsonArray areasOfInterest = (JsonArray) (reader.read());
					int numAreasOfInterest = areasOfInterest.size();

					// Lists to hold the data of each type; separated into lists
					// by region of interest. (Eg, all of the faces from the
					// first region of interest would be in faceData.get(0).)
					// This would be typed as List<List<Blob>> but it doesn't
					// compile that way.
					List<ArrayList<Blob>> shellData = new ArrayList<ArrayList<Blob>>(numAreasOfInterest);
					List<ArrayList<Blob>> faceData = new ArrayList<ArrayList<Blob>>(numAreasOfInterest);

					// Whether or not we actually have blob information.
					boolean haveShells = false;
					boolean haveFaces = false;

					// Each are of interest is an object with keys for "Faces"
					// and "Shells". We would use the original JsonArrays, but
					// we need to convert the data into Blob objects.
					for (int i = 0; i < numAreasOfInterest; i++) {
						JsonObject aoi = areasOfInterest.getJsonObject(i);
						ArrayList<Blob> current = new ArrayList<Blob>();

						JsonArray shells = aoi.getJsonArray("Shells");
						shellData.add(current);
						for (JsonValue val : shells) {
							haveShells = true;
							JsonArray shell = (JsonArray) (val);
							current.add(parseEventDataToBlob(shell, i /* roi */, CVEvent.EVENT_TYPE.SHELL));
						}

						JsonArray faces = aoi.getJsonArray("Faces");
						current = new ArrayList<Blob>();
						faceData.add(current);
						for (JsonValue val : faces) {
							haveFaces = true;
							JsonArray face = (JsonArray) (val);
							current.add(parseEventDataToBlob(face, i /* roi */, CVEvent.EVENT_TYPE.FACE));
						}

					}

					// Dispatch the events with the associated data.
					eventDispatcher(faceData, haveFaces, shellData, haveShells, numAreasOfInterest);

				} catch (ClassCastException e) {
					System.out.println("Unexpected message format from server: " + e);
				}
			}

			out.close();
			in.close();
			mySocket.close();
		} catch (UnknownHostException e) {
			System.err.println("Don't know about host: localhost.");
			System.out.println("Can't find localhost");
			System.exit(1);
		} catch (IOException e) {
			System.err.println("Couldn't get I/O for the connection to: " + hostname + " on port " + port);
			// undecided whether to have the system exit if only one of the
			// connections is unsuccessful
			// System.exit(1);

		}

	}

	/**
	 * Checks the number of each type of blob and dispatches the appropriate
	 * events. If there is at least one of a type of blob, that type of event is
	 * dispatched. If there is at least one blob, then ALL_BLOB event type is
	 * dispatched.
	 * 
	 * @param faceData
	 *            The list of face events, separated into lists by region of
	 *            interest
	 * @param haveFaces
	 *            Whether or not any faces were detected
	 * @param shellData
	 *            The list of shell events, separated into lists by region of
	 *            interest
	 * @param haveShells
	 *            Whether or not any shells were detected
	 * @param numAreasOfInterest
	 *            The number of regions of interest we currently have
	 */
	private void eventDispatcher(List<ArrayList<Blob>> faceData, boolean haveFaces, List<ArrayList<Blob>> shellData,
			boolean haveShells, int numAreasOfInterest) {
		CVEventData cvEventData = new CVEventData(faceData, shellData, numAreasOfInterest);

		if (haveShells) {
			dispatchEvent(new CVEvent(CVEvent.EVENT_TYPE.SHELL, cvEventData));
		}
		if (haveFaces) {
			dispatchEvent(new CVEvent(CVEvent.EVENT_TYPE.FACE, cvEventData));
		}
		// If we have either type of blob (or both), send out a general blob
		// event.
		if (haveShells || haveFaces) {
			dispatchEvent(new CVEvent(CVEvent.EVENT_TYPE.ALL_BLOBS, cvEventData));
		}
	}

	/**
	 * Each listener occupies two elements - the first is the listener class and
	 * the second is the listener instance
	 * 
	 * @param evt
	 *            The event to dispatch
	 */
	private void dispatchEvent(CVEvent evt) {
		Object[] listeners = listenerList.getListenerList();
		// each listener is going to be held in two elements
		for (int i = 0; i < listeners.length; i += 2) {
			if (listeners[i] == CVEventListener.class) {
				if (evt.getType().equals(CVEvent.EVENT_TYPE.SHELL)) {
					((CVEventListener) listeners[i + 1]).shellsArrived(evt);
				} else if (evt.getType().equals(CVEvent.EVENT_TYPE.FACE)) {
					((CVEventListener) listeners[i + 1]).facesArrived(evt);
				} else if (evt.getType().equals(CVEvent.EVENT_TYPE.ALL_BLOBS)) {
					((CVEventListener) listeners[i + 1]).blobsArrived(evt);
				}
			}
		}
	}

	/**
	 * Parses the event data to find the x, y, width, and height of the current
	 * event.
	 *
	 * @param blobData
	 *            A JsonArray containing the x, y, width, and height of a blob
	 * @param roi
	 *            The region of interest this blob occurred in
	 * @param type
	 *            The type of event (shell or face)
	 * @return A blob object with the x, y, width, and height that was parsed
	 *
	 **/
	private Blob parseEventDataToBlob(JsonArray blobData, int roi, CVEvent.EVENT_TYPE type) {
		// Get the leftX, topY, width, and height from the JSON data.
		double xCoord = blobData.getJsonNumber(0).doubleValue();
		double yCoord = blobData.getJsonNumber(1).doubleValue();
		double width = blobData.getJsonNumber(2).doubleValue();
		double height = blobData.getJsonNumber(3).doubleValue();

		return new Blob(type, xCoord, yCoord, width, height, roi);
	}

}

File: client/library/python/ihart/cv_manager.py
"""
Connects to the server to receive and parse information about events.
"""

import json # for message parsing
import select # for handling sockets
import socket # for communicating with the server

from blob import Blob
from cv_event_data import CVEventData


class CVManager:
    """
    Sets up a non-blocking connection to the server via sockets. New events can
    be retrieved via the getNewEvents() method.
    """

    _client_socket = None
    _connectionSucceeded = False

    def __init__(self, host="localhost", port=5204):
        self._client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

        try:
            self._client_socket.connect((host, port))

            # Make the socket non-blocking
            # (see http://docs.python.org/library/socket.html#socket.socket.setblocking)
            self._client_socket.setblocking(0)

            self._connectionSucceeded = True

        # If we can't connect, report the error; we can't really recover from
        # not being able to establish a connection, but we don't want to crash.
        except socket.error, e:
            print ">>> Unable to establish a connection with the iHart server:"
            print e


    def getNewEvents(self):
        """
        @return a list of CVEventData objects containing information about faces and
            motion (aka shells) detected by the server. The list will be in order the
            events arrived. The list will be empty if there was nothing to detect.
            None will be returned if the server sent no data, or if an error occurred.
        """
        # We can't do anything if the connection wasn't established
        if not self._connectionSucceeded:
            return None

        # Try to grab a message from the server and parse it into an event data
        # object.
        message = self._tryToReceiveEvents()
        if message:
            return self._parseMessage(message)
        else:
            return None


    def destroy(self):
        """
        Close the connection to the server.
        @return None
        """
        self._client_socket.close()


    def _tryToReceiveEvents(self):
        """
        If there is anything available from the server, grab it and parse it.

        See http://stackoverflow.com/questions/8386679/why-am-i-receiving-a-string-from-the-socket-until-the-n-newline-escape-sequence
        for using select with non-blocking sockets.

        @return a message sent from the server containing detected faces and
            motion (shells), or None if there was nothing to detect (or an
            error ocurred)
        """
        # See if the client socket is ready to be read from.
        read_ready, _, _ = select.select([self._client_socket], [], [])

        if self._client_socket in read_ready:
            # The socket has data ready to be received
            message = ""
            continue_recv = True
            while continue_recv:
                try:
                    # Try to receive som data
                    message += self._client_socket.recv(1024)
                except socket.error, e:
                    if e.errno != socket.errno.EWOULDBLOCK:
                        # Error! Print it and tell main loop to stop
                        print 'Error: %r' % e
                    # If e.errno is errno.EWOULDBLOCK, then no more data
                    continue_recv = False

            if message:
                return message

        # If we get to this point and haven't returned anything, there's
        # nothing to return (either socket wasn't ready or there was an error).
        return None


    def _parseMessage(self, message):
        """
        Parse a JSON message from the server. See createInformationString in
        http://ihart-mhc.github.io/software/data.html for details on the format
        of the message.
        """
        # Remove characters that the json parser doesn't like
        message = message.translate(None, "\0\r")

        # We might have received multiple event objects, which are separated by
        # newline characters.
        messages = message.split("\n")

        # Process each message into an object containing all of the event
        # information.
        events = []
        for message in messages:
            if not message:
                continue

            # TODO: wrap this in a try-except block in case the parsing fails?
            data = json.loads(message)
            numRoi = len(data)

            facesByRegion = []
            shellsByRegion = []

            # Process the blobs in each region of interest and store them in the
            # corresponding list.
            for roi, blobs in enumerate(data):
                facesByRegion.append([ Blob(b[0], b[1], b[2], b[3], roi, Blob.FACE) for b in blobs["Faces"] ])
                shellsByRegion.append([ Blob(b[0], b[1], b[2], b[3], roi, Blob.SHELL) for b in blobs["Shells"] ])

            events.append(CVEventData(facesByRegion, shellsByRegion, numRoi))

        return events




# For testing/development
if __name__ == "__main__":
    cvm = CVManager()
    for i in range(10):
        print cvm.getNewEvents()
    cvm.destroy()
